src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -201,135 +210,174 @@ private Type EmitCoercionCode(AssociativeNode arg, Type argType, Type param)
+                                    p => p.Name == nameof(IEnumerator<Source>.Current)).FirstOrDefault();
             mInfo = prop.GetAccessors().FirstOrDefault();
-
             EmitOpCode(OpCodes.Callvirt, mInfo);
 
-            // unboxing
-            if (typeof(T) == typeof(double))
+            localBuilder = DeclareLocal(typeof(Source), "element in source array");
+            var sourceArrayElementIndex = localBuilder.LocalIndex;
+            EmitOpCode(OpCodes.Stloc, sourceArrayElementIndex);
+
+            // target[i] = source_element;
+            EmitOpCode(OpCodes.Ldloc, newArrIndex);
+            EmitOpCode(OpCodes.Ldloc, counterIndex);
+            EmitOpCode(OpCodes.Ldloc, sourceArrayElementIndex);
+            if (typeof(Target) == typeof(double))
             {
-                EmitOpCode(OpCodes.Call, typeof(Convert).GetMethod(nameof(Convert.ToDouble), new Type[] { typeof(object) }));
+                EmitOpCode(OpCodes.Conv_R8);
                 EmitOpCode(OpCodes.Stelem_R8);
             }
-            else if (typeof(T) == typeof(int))
+            else if (typeof(Target) == typeof(int))
             {
-                EmitOpCode(OpCodes.Call, typeof(Convert).GetMethod(nameof(Convert.ToInt32), new Type[] { typeof(object) }));
+                EmitOpCode(OpCodes.Conv_I4);
                 EmitOpCode(OpCodes.Stelem_I4);
             }
-            else if (typeof(T) == typeof(long))
+            else if (typeof(Target) == typeof(long))
             {
-                EmitOpCode(OpCodes.Call, typeof(Convert).GetMethod(nameof(Convert.ToInt64), new Type[] { typeof(object) }));
+                EmitOpCode(OpCodes.Conv_I8);
                 EmitOpCode(OpCodes.Stelem_I8);
             }
             // i++;
             EmitOpCode(OpCodes.Ldloc, counterIndex);
             EmitOpCode(OpCodes.Ldc_I4_1);
-
             EmitOpCode(OpCodes.Add);
-
             EmitOpCode(OpCodes.Stloc, counterIndex);
 
-            // i < len;
-            MarkLabel(loopCondLabel.Value,"label:cond");
+            MarkLabel(loopCondLabel.Value, "label:loop condition");
+            EmitOpCode(OpCodes.Ldloc, enumeratorIndex);
 
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
+            mInfo = typeof(IEnumerator).GetMethod(nameof(IEnumerator.MoveNext));
+            EmitOpCode(OpCodes.Callvirt, mInfo);
 
-            EmitOpCode(OpCodes.Ldloc, sourceArrayLengthIndex);
+            var leaveLabel = DefineLabel();
+            EmitOpCode(OpCodes.Brtrue_S, loopBodyLabel.Value);
+            EmitOpCode(OpCodes.Leave_S, leaveLabel.Value);
 
-            EmitOpCode(OpCodes.Clt);
+            EmitOpCode(OpCodes.Ldloc, enumeratorIndex);
 
-            EmitOpCode(OpCodes.Brtrue_S, loopBodyLabel.Value);
+            var finallyLabel = DefineLabel();
+            EmitOpCode(OpCodes.Brfalse_S, finallyLabel.Value);
+
+            EmitOpCode(OpCodes.Ldloc, enumeratorIndex);
+
+            mInfo = typeof(IDisposable).GetMethod(nameof(IDisposable.Dispose));
```

Comment:
what is going on here? What needs to be disposed during coercion?

Reply:
This code is emitting the following `foreach ` loop and we need to dispose the enumerator used in the loop.
```
        var len = a.Count();    // where a is an IEnumerable<Source>
	var res = new Target[len];
	int c = 0;
	foreach(var i in a)
	{
		res[c] = i;
		c++;
	}
```